### PR TITLE
chore(networkmonitor): tool to discover and provide metrics on peers

### LINKED ISSUE #1010: Waku v2 monitoring node
## Problem

We need a node that can be deployed to a specific Waku v2 network that can be used to gather various network metrics and monitor the overall health of the network. Note that this should not be confused with the ["Waku v2 canary service"](https://github.com/status-im/nwaku/issues/754) which will function more like a tool that can be used for checking the health of specific nodes on an ad-hoc basis.

## Suggestion

This node could, for example,
- "walk" the discv5 DHT in the network to get an idea of the number of nodes participating in the network
- try to get an idea of the number of unique peer IDs seen in the network (using various discovery methods)
- keep track of the number of messages per:
  - protocol
  - content topic
  - application (first part of content topic)

## Tracking:

- [x] metrics on discovered and connected peers: https://github.com/status-im/nwaku/pull/1290
- [x] metrics on messages (application, topic, etc): #1335 
- [ ] run an instance of `networkmonitor` exposing its data in a grafana dashboard.
  - [x] https://github.com/status-im/infra-misc/issues/104#issuecomment-1315622948
  - [ ] https://github.com/status-im/nwaku/pull/1401

### PR DESCRIPTION
**Pull request by alrevuelta**
Related to #1010

Description:
* Adds a tool that constantly tries to discover new peers in the network.
* It also tries to connect to them.
* See `README.md` for instructions and available metrics.
* Goal is to run a long lived instante of this tool, with a grafana dashboard we can monitor.
* Some metrics are exposed to prometheus to be fetched. Labels are constrained according do @jakubgs feedback.
* Other metrics, related to individual peers information are exposed directly via a rest api, ready to be accessed by i.e. Grafana. Note that these metrics are not exposed to prometheus.

Limitations:
* Naive implementation, with all data stored in memory. If the network scales to thousands of node the performance might be an issue.
* Routing table is not emptied.

Usage:

```console
$ make networkmonitor
$ ./build/networkmonitor --log-level=INFO --b="enr:-Nm4QOdTOKZJKTUUZ4O_W932CXIET-M9NamewDnL78P5u9DOGnZlK0JFZ4k0inkfe6iY-0JAaJVovZXc575VV3njeiABgmlkgnY0gmlwhAjS3ueKbXVsdGlhZGRyc7g6ADg2MW5vZGUtMDEuYWMtY24taG9uZ2tvbmctYy53YWt1djIucHJvZC5zdGF0dXNpbS5uZXQGH0DeA4lzZWNwMjU2azGhAo0C-VvfgHiXrxZi3umDiooXMGY9FvYj5_d1Q4EeS7eyg3RjcIJ2X4N1ZHCCIyiFd2FrdTIP"
```

And:
* See prometheus metrics `http://localhost:8008/metrics`
* See custom metrics `http://localhost:8009/allpeersinfo`

### CODE REVIEW

File: tools/networkmonitor/networkmonitor.nim
Comment (LNSD) [inline tools/networkmonitor/networkmonitor.nim:22]:
```suggestion
import
  ../../waku/v2/protocol/waku_message,
  ../../waku/v2/node/discv5/waku_discv5,
  ../../waku/v2/node/wakunode2,
  ../../waku/v2/utils/wakuenr,
  ./networkmonitor_metrics,
  ./networkmonitor_config
```

**Review by jakubgs — CHANGES_REQUESTED** [review]
You cannot put unconstrained values like ENRs in labels for Prometheus metrics. This will lead to an inevitable explision in cardinality which will kill the performance of any Prometheus instance or remote write backend:

>__CAUTION__: Remember that every unique combination of key-value label pairs represents a new time series, which can dramatically increase the amount of data stored. Do not use labels to store dimensions with high cardinality (many different label values), such as user IDs, email addresses, or other unbounded sets of values.

https://prometheus.io/docs/practices/naming/#labels

The only way around that is using some kind of limit set of buckets.

**Comment by alrevuelta** [comment]
Expanding a bit on the issue brought up by @jakubgs.

## Problem:

I want to have strings as metrics, something not trivial, but can be achieved with `labels`.

```nim
declarePublicGauge networkmonitor_example, "Description", labels = ["peerId", "info"]
```

We update it with a dummy `0.0` value, ok.
```nim
networkmonitor_example.set(0.0, labelValues = ["peerId1", "ip1,enr1,wakuflags1,lastconnection1"])
networkmonitor_example.set(0.0, labelValues = ["peerId2", "ip2,enr2,wakuflags2,lastconnection2"])
```

But if we modify one label `lastconnection2` value, we will have another time series point. Something that can break prometheus at some point.

```nim
networkmonitor_example.set(0.0, labelValues = ["peerId1", "ip1,enr1,wakuflags1,lastconnection2"])
```

Proof of it, `peerId1` has now two datapoints, and we only need one.

```console
networkmonitor_example{peerId="peerId1",info="ip1,enr1,wakuflags1,lastconnection1"} 0.0
networkmonitor_example_created{peerId="peerId1",info="ip1,enr1,wakuflags1,lastconnection1"} 1666957272.0
networkmonitor_example{peerId="peerId2",info="ip2,enr2,wakuflags2,lastconnection2"} 0.0
networkmonitor_example_created{peerId="peerId2",info="ip2,enr2,wakuflags2,lastconnection2"} 1666957272.0

networkmonitor_example{peerId="peerId1",info="ip1,enr1,wakuflags1,lastconnection2"} 0.0
networkmonitor_example_created{peerId="peerId1",info="ip1,enr1,wakuflags1,lastconnection2"} 1666957272.0
```

## Solution
As suggested by Jakub I tried using a limited set of buckets, but since I need strings I had to use `labels` having the same issue as the beginning. However I found an alternative that should fix the problem of having infinite time-series points.

Going into `metrics` tables, its possible to update an existing `key` if it was already present or create a new if not. With this we no longer have an infinite increasing number of time series points.

```nim
proc addOrUpdate(peerId: string, content: string, metricGauge: Gauge) = 
  for i in networkmonitor_example.metrics.keys:
    if i[0] == peerId:
      echo "already there, update it"
      metricGauge.metrics[@[i[0], i[1]]][0].labelValues[1] = content
      return
  echo "not there, add new field"
  metricGauge.set(0.0, labelValues = [peerId, content])
```

Long story short, it's possible to use have a `key,value` like set of metrics using labels with a constrained number of time series points.

```nim
addOrUpdate("peerId2", "ip1,enr1,wakuflags1,lastconnection2", networkmonitor_example)
```

@jakubgs :
* Do you think this solves the issue you brought up? I have verified it by looking into localhost:8008/metrics, observing that no new points are added with new label combinations. That should be enough.
* There will be ofc 1 entry per active peer in the network. Luckily this is around 10 and will take time to reach levels where prometheus can't handle it. I recall you mentioned few hundreds. If you think its safer I can add a threshold of `x` entries, and after that just error and drop values.
* Note that this is temporal. If waku network grows to thousands of nodes, we will ofc have a proper DB.

**Comment by jakubgs** [comment]
No, this does not solve the issue as peer ids are unconstrained. Also not sure why you think you can constrain the number of labels created based on runtime data of a service that can stop and start. Every restart will make it forget what labels it has already created, in effect nullifying any attempt at constraining the number of values.

The only solutions are:
- Not use any unconstrained values for labels like peer IDs at all.
- Use a specific constrained set of buckets for values of labels that cannot grow.
- Simply not use Prometheus as backend for data it was not intended for.

**Comment by alrevuelta** [comment]
Thanks for the input @jakubgs. Perhaps Prometheus is overkill for this use case, so would go with "Simply not use Prometheus as backend for data it was not intended for." I have found "Grafana Infinity Datasource" plugin that allows me to get all the info I want to display via http get. So I will connect grafana directly with this tool (networkmonitor) to display this data without prometheus in between. Anything against that?

**Comment by rymnc** [comment]
Do we *really* need the location information?

**Comment by alrevuelta** [comment]
@rymnc 
> Do we really need the location information?

Anyone can see it, so why not? I think it's interesting to see how decentralized the network is in terms of location.

**Comment by alrevuelta** [comment]
@LNSD Yes, it is ready for review, that's why I requested it to be reviewed. There are some TODOs but will be addressed in other PRs.

Regarding #1335 I will merge it to master once this one is merged, but it's based on this PR for a clearer diff. Note that #1335 is also ready for review.

File: tools/networkmonitor/networkmonitor_utils.nim
Comment (jm-clius) [inline tools/networkmonitor/networkmonitor_utils.nim:29]: If I understand correctly this will always be called for each discovered peer? I'm wondering what happens when the network scales to 1000s of nodes (here's hoping)? Would it make sense (perhaps only in a future PR) to split the monitoring into three different reporting procedures, spawned independently, that reports metrics on (1) what can be determined from discovered ENRs directly (no wait) (2) ipToLocation (with a period that respects the API limits) (3) connect to those nodes and report supported protocols? In other words the discovery loop (faster) is somewhat decoupled from the slower connection and location lookup procs?

**Review by jm-clius — COMMENTED** [review]
This is a very useful tool! Thanks for thinking creatively about what we'd need here. I've added some comments below, mostly related to separating some of the logical components for future extensibility if this is to become a more general tool (such as reporting more capabilities of nodes once we have a capability discovery protocol). If suggestions are unclear, feel free to ping me to clarify. Also happy if some parts are left for future PRs.

File: tools/networkmonitor/networkmonitor.nim
Comment (alrevuelta) [reply tools/networkmonitor/networkmonitor.nim:164]: yep, this difference advertised (as per enr) and supported (as per protocols after a connection) exists.

this todo related to a refactor im preparing. nothing different, same functionality. will followup in another pr, as its unrelated to this (but can simplify some things)

File: tools/networkmonitor/networkmonitor.nim
Comment (alrevuelta) [reply tools/networkmonitor/networkmonitor.nim:203]: sure, both comments are fair, will refactor that.

Regarding the use of timers, i have some concerns. imho using setTimer() to schedule a recurrent task as done in [here](https://github.com/status-im/nwaku/blob/master/waku/v2/node/waku_metrics.nim#L74-L76) seems convoluted to me and a bit hackish. I understand you are asking to do the same here. My reasons to think that:

* on each iteration you schedule again and again the task, x time from now, some kind of recursive setTimer.
* you need an external timer to trigger the function the first time and then a second internal timer that is called, again and again.
* this adds extra code and its more difficult to reason.
* in general the amount of code is more than simply `while true:` and `sleepAsync` with an `asyncSpawn`.
* seems hackish as we are constantly scheduling the task x time from now.
* i don't see nim-libp2p nor nimbus using `setTimer` for scheduling recurrent tasks.
* [nim-libpb pubsub loop](https://github.com/status-im/nim-libp2p/blob/fa5d102370cff4e68b47ac633153b22609b8aa41/libp2p/daemon/daemonapi.nim#L1286) follows the pattern i suggest. [same in this example](https://github.com/status-im/nim-libp2p/blob/fa5d102370cff4e68b47ac633153b22609b8aa41/examples/go-daemon/bootstrap.nim)
* same for [nimbus](https://github.com/status-im/nimbus-eth2/blob/97a1cdc4902eac0149b0e2f8a0ac9fa663afe408/beacon_chain/nimbus_beacon_node.nim#L1386)

As far as I know, we don't have a `ScheduleAtFixedRate(func, time)` like in other languages, so imho the best way would be to follow what nimlibp2p/nimbus does.

```nim
proc recurrentTask() {.async.} = 
  echo "do stuff"
  sleepAsync(5000)
```

and somewhere else 
```nim
asyncSpawn recurrentTask()
```

@LNSD thoughts?

File: tools/networkmonitor/networkmonitor_utils.nim
Comment (alrevuelta) [reply tools/networkmonitor/networkmonitor_utils.nim:29]: yup, definitely. i had something similar in mind :)

**Comment by alrevuelta** [comment]
@jm-clius just fixed the comments! with an emphasis on having a clearer `main`, easier to read with high-level procedures on what's being done. start this, start that, craw this, blabla.

Also added the `.push raises:` that was missing, took me some time to fix everything :)

Note that [i would prefer to not user timers](https://github.com/status-im/nwaku/pull/1290#discussion_r1013231426) for recurrent tasks, as looks like they are not meant for that.

File: tools/networkmonitor/networkmonitor.nim
Comment (LNSD) [reply tools/networkmonitor/networkmonitor.nim:151]: I'm afraid I disagree with your rationale about the timer tasks. But I will base my reasoning on facts instead of opinions. 

Short answer and a rule of thumb: 
> Using a `sleep` is **always** a hack unless it is strictly unavoidable. Some exceptions could be building a task scheduler, an OS, or working on bare metal system programming.

The fact that _nimbus_ and _nim-libp2p_ don't do it could be explained by other reasons we don't know. For example, amongst other things, `nim-libp2p` uses exception, and the status nim style guide [recommends the opposite](https://status-im.github.io/nim-style-guide/errors.exceptions.html#:~:text=In%20general%2C%20prefer%20explicit%20error%20handling%20mechanisms.).

So nothing done in other projects should be taken as written in stone.

--------------

### Some context

Checking the implementation of `asyncSleep()` it registers a timer in the `nim-chonos` scheduler:

https://github.com/status-im/nim-chronos/blob/6525f4ce1d1a7eba146e5f1a53f6f105077ae686/chronos/asyncloop.nim#L983-L1000

Also, as @rymnc pointed out in [this PR](https://github.com/status-im/nwaku/pull/1094#discussion_r957177210), there is a (stale) issue in `nim-chronos` that will bring (or not 🤷🏼) the desired interval timer primitive to the party like other technologies have (e.g., NodeJS) given that it is a common task: https://github.com/status-im/nim-chronos/pull/62

### Both approaches comparison

Let's compare the _while/sleep approach_ and the _interval/set timer_ approaches:

The **nested timers approach**, depending on how you implement it, can make a repetitive task start to be deterministically executed, i.e., if, at the moment of executing the callback, the first thing you do is to re-schedule the following execution, you can be sure that it will be executed next time at t+<interval you configured> (of course if the `nim-chronos`' dispatcher is not busy with any blocking workload), independently of the duration of that callback's task duration. This approach can also make the time separation between task executions deterministic by re-scheduling the task after the task is executed.

On the other hand, the **`while true` and sleep** approach: Requires to be within an async method that should be spawned by `asyncSpawn`. And the start of the execution is not deterministic. What is deterministic is the spacing between the different task executions; i.e., if your task takes X to be executed, then the next task will be executed at X+sleep time, and the following at X+sleep time+X'+sleep time (NB that we are working with IO so the X time can vary significantly).

So the first approach is objectively more versatile and covers scenarios that the second does not cover out of the box.

-------------

I hope this makes it clear why.

Writing this down takes a lot of effort. I know that it is helpful to understand the reasons for our suggestions, but it is time-consuming and hinders other team members' productivity. When someone suggests something, I start thinking, "why?" and look for the answers/facts to understand it.

I value your opinion, but we are doing engineering, which should be based on facts. And also, please, be mindful of others' time.

**Review by LNSD — COMMENTED** [review]
Please, check the comments I have written so far.

I have not finished reviewing the `main()` function. I need to jump on something else. I will continue reviewing it later.

File: tools/networkmonitor/networkmonitor.nim
Comment (alrevuelta) [reply tools/networkmonitor/networkmonitor.nim:221]: Mind elaborating? I was following [wakunode2](https://github.com/status-im/nwaku/blob/master/apps/wakunode2/wakunode2.nim#L533-L548). Isn't it the same?

File: tools/networkmonitor/networkmonitor.nim
Comment (alrevuelta) [reply tools/networkmonitor/networkmonitor.nim:223]: I have a strong opinion on this. CLI flags should be logged with info, as they are one of the most important things when your startup a program.

File: tools/networkmonitor/networkmonitor.nim
Comment (alrevuelta) [reply tools/networkmonitor/networkmonitor.nim:151]: Thanks for the heads up about minding others' time, but it's up to each one of us to allocate time wisely, so keep that in mind.

While I agree with your "fact-based arguments" I don't agree with your rule of thumb. Engineering is also pragmatic, so since i) I don't need deterministic time spacing and ii) `sleepAsync`/`asyncSpawn` is simpler and less convoluted than timers, that should be sufficient for this PR.

For sure I will consider using [interval](https://github.com/status-im/nim-chronos/pull/62) if merged and if I need a deterministic time spacing I will use your solutions with timers.

Thanks for the detailed explanation.

File: tools/networkmonitor/networkmonitor_metrics.nim
Comment (alrevuelta) [reply tools/networkmonitor/networkmonitor_metrics.nim:75]: sure, good catch, missed it in the refactoring.
i'm confused since with `{.push raises: [].}` this shouldn't compile, or? as its missing the `{.raises: [xxx].}`

File: tools/networkmonitor/networkmonitor_metrics.nim
Comment (alrevuelta) [reply tools/networkmonitor/networkmonitor_metrics.nim]: sure, fixed in 4de12f2a226993d2ca7b1cca3131bb22957bfc79

File: tools/networkmonitor/networkmonitor_metrics.nim
Comment (alrevuelta) [reply tools/networkmonitor/networkmonitor_metrics.nim]: good catch thanks. 4de12f2a226993d2ca7b1cca3131bb22957bfc79

File: tools/networkmonitor/networkmonitor_metrics.nim
Comment (alrevuelta) [reply tools/networkmonitor/networkmonitor_metrics.nim]: sure, 4de12f2a226993d2ca7b1cca3131bb22957bfc79

File: tools/networkmonitor/networkmonitor_config.nim
Comment (alrevuelta) [reply tools/networkmonitor/networkmonitor_config.nim:11]: added in 4de12f2a226993d2ca7b1cca3131bb22957bfc79

File: tools/networkmonitor/networkmonitor_config.nim
Comment (alrevuelta) [reply tools/networkmonitor/networkmonitor_config.nim]: fix 4de12f2a226993d2ca7b1cca3131bb22957bfc79

File: tools/networkmonitor/networkmonitor.nim
Comment (LNSD) [reply tools/networkmonitor/networkmonitor.nim:221]: No, it's not the same.

https://github.com/status-im/nwaku/blob/master/apps/wakunode2/config.nim#L521-L537

File: tools/networkmonitor/networkmonitor.nim
Comment (LNSD) [reply tools/networkmonitor/networkmonitor.nim:228]: And I disagree on the logging level. Hence my code suggestion.

File: tools/networkmonitor/networkmonitor.nim
Comment (LNSD) [reply tools/networkmonitor/networkmonitor.nim:221]: The comment is no longer necessary, given that we are wrapping the load proc with a try/catch and returning a result.

**Comment by alrevuelta** [comment]
@LNSD Fixed all the comments. Is there anything else you want to address? No pressure if you need more time for review, just want to make sure its not blocked for no reason.

File: waku.nimble
Comment (LNSD) [inline waku.nimble:114]: The `-d:chronicles_runtime_filtering:on"` should be specified in the `nim.cfg` next to the tool/app binary. Check the other binaries' `nim.cfg` for reference

**Review by LNSD — COMMENTED** [review]
Given the situation and my POV, merge it if you think it is what you should do

File: waku.nimble
Comment (alrevuelta) [reply waku.nimble]: sure, added it in 725cd9f7c41b502f0c21f6b0a3f0f3c110c72f83

The thing I don't like is that you don't get in the compiling logs the flags you are using in `nim.cfg`, which imho can lead to problems. Increasing `verbosity` the max I managed to see was this. used [this](https://nim-lang.org/docs/nimc.html#compiler-usage-configuration-files) as reference.
```console
Hint: used config file '/Users/alrevuelta/Github/nwaku/vendor/nimbus-build-system/vendor/Nim/config/nim.cfg' [Conf]
Hint: used config file '/Users/alrevuelta/Github/nwaku/vendor/nimbus-build-system/vendor/Nim/config/config.nims' [Conf]
Hint: used config file '/Users/alrevuelta/Github/nwaku/config.nims' [Conf]
Hint: used config file '/Users/alrevuelta/Github/nwaku/tools/networkmonitor/nim.cfg' [Conf]
```

Unless I'm missing something `chronicles_runtime_filtering` is not present in the logs while being a flag, which imho is dangerous. Since I like being explicity, I prefered to have it directly in the makefile.

But note that I adapted the code to your suggestion following the repo "pattern".

**Comment by alrevuelta** [comment]
since `test2-ubuntu-latest` is ok, bypassing branch protection due to a known issue with `test2-macos-latest`

**Pull request merged**
